### PR TITLE
Feature/adds development container

### PR DESCRIPTION
**Pull request by jasonwashburn**
This PR adds config and scripts for a devcontainer running on the rust/bullseye variant. The WASI-SDK and Spin versions used at build time, can be modified by changing args passed to Dockerfile in devcontainer.json. 

Compilation of cPython and spin-python-sdk are performed in a post creation command script since they rely on the repository being available (not the case during build time when using on GitHub codespaces)

I've tested it successfully as is on both GitHub Codespaces, and on an x86 linux host, but it could probably use a little more testing in other configs, ie: on windows, or maybe with different versions of wasi-sdk / spin version, though I don't know the currently compatibility story there, so maybe not relevant.

Known issues:
- Doesn't currently work on newer (m1, m2) MacOS hosts due to glibc2 issues with spin installer on aarch64 bullseye. (believe related to [spin/168](https://github.com/fermyon/spin/issues/168) and similar to [bartholomew/150](https://github.com/fermyon/bartholomew/issues/150)
- Based on previous experience locally, I also expect to have trouble with compiling cPython on newer MacOS hosts as well once past the spin installer

Opening up draft PR to facilitate discussion and collect feedback on experience, or thoughts on improvements to the config, extensions, included python tooling etc.

### CODE REVIEW

**Comment by jasonwashburn** [comment]
It seems there's an issue with trying to switch the variant to buster. When you do so, the spin install script fails for a missing version of glibc (See output below). Looks to be related to https://github.com/fermyon/spin/issues/168 and https://github.com/fermyon/bartholomew/issues/150 as well, not necessarily an issue with the python sdk or the devcontainer, other than maybe don't use buster.

reproducible by building the devcontainer manually:
`docker build --build-arg VARIANT=buster .devcontainer/.`

```plaintext
Step 10/10 : RUN cd /tmp     && curl -fsSL https://developer.fermyon.com/downloads/install.sh | bash -s -- -v $SPIN_VERSION     && sudo mv spin /usr/local/bin/     && cd -
 ---> Running in 54c0fee46bb9
Step 1: Downloading: https://github.com/fermyon/spin/releases/download/canary/spin-canary-linux-amd64.tar.gz
Done...

Step 2: Decompressing: spin-canary-linux-amd64.tar.gz
crt.pem
spin.sig
README.md
LICENSE
spin
./spin: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by ./spin)
./spin: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.29' not found (required by ./spin)
The command '/bin/sh -c cd /tmp     && curl -fsSL https://developer.fermyon.com/downloads/install.sh | bash -s -- -v $SPIN_VERSION     && sudo mv spin /usr/local/bin/     && cd -' returned a non-zero code: 1
```

**Comment by mikkelhegn** [comment]
Does installing the build dependencies, like described here: https://developer.fermyon.com/spin/install#building-spin-from-source help?

**Comment by jasonwashburn** [comment]
I hadn't tried it, but I suspect it probably would since it should link against the correct glibc. Wasn't sure of the best way to go on it. Since it's being used for development, my initial thought was to just duplicate the environment that the main spin repo is build in, to keep things the same. Though, it wouldn't be linked, so there's just a much a chance of getting out of sync, so maybe building from scratch based on the selected variant and spin version. 🤔

**Comment by jasonwashburn** [comment]
hmm...I had thought I responded to that last question already, I guess not. Long story short, switching to building spin from source didn't fix the issue, I started running into the same glibc error messages from the wasi sdk as well. After taking another look at the intent of a devcontainer, to get a dev environment with dependencies up and running quickly and reproducibly, I'm going to limit the devcontainer install/build to the prerequisite installed dependencies listed in the README and get a user right up to the point where they can run the build scripts (build-python.sh and make) to match the instructions in the README, rather automating the full build when the container starts up. I think this will work better with future changes that will require changes to the build system, cpython versions etc.

**Comment by jasonwashburn** [comment]
Alright, I think I have it in a good place. Tested it on Ubuntu, Ubuntu in WSL , and on Github Codespaces by pulling the repo from scratch, building the devcontainer in vscode and then following the instructions in the readme to build the plugin, then test it with the example hello world app. I made an update to the readme about needing to do a git submodule update before trying to run the build-python.sh.

File: .devcontainer/postCreateCommands.sh
Comment (mikkelhegn) [inline .devcontainer/postCreateCommands.sh:1]: Is this file still needed?

File: README.md
Comment (mikkelhegn) [inline README.md:44]: Could this and the CPython build be part of the container creation?

**Review by mikkelhegn — CHANGES_REQUESTED** [review]

File: README.md
Comment (jasonwashburn) [reply README.md:44]: We could actually do a couple of things here. I flipped back and forth a couple of times on it previously. It all comes down to the usual workflow...if there is rarely a need to rebuild cpython or make changes to the branch etc, then it probably makes sense to move it into the dockerfile and take advantage of image caching...the build is already long enough to be a "walk away for a bit" build the first time, so I'm not adverse to just including the python compile as well.

Ultimately I made the decision to leave it this way after looking at how the development steps are listed in the README...ie: install these dependencies (stuff I put in dockerfile) and then steps begin with "run this script" which I left up to the user. 

I'll defer to @dicej and the guys at Fermyon though as far as which way would be a better developer experience given the normal workflow and tasks that are often performed?

File: .devcontainer/Dockerfile
Comment (jasonwashburn) [reply .devcontainer/Dockerfile:38]: ...as we discussed briefly on discord, so far, successful M1/M2 builds have eluded me...I'm going to try the ubuntu build suggestion from @dicej and hopefully we'll be good...failing that, I agree with your suggestion to fail fast on the build if the user is building on aarch64 rather than  putting them through the long build/compile process.

File: README.md
Comment (jasonwashburn) [reply README.md]: As discussed on discord, I switch the devcontainer to use an artifact compiled version of wasi cpython, so running these commands is no longer necessary when using it. Added notes to the affected steps in the README.

File: .devcontainer/Dockerfile
Comment (jasonwashburn) [reply .devcontainer/Dockerfile:24]: Should be working now, I was able to launch the devcontainer and run the example using my m2 macbook pro.

File: .devcontainer/onCreateCommands.sh
Comment (mikkelhegn) [inline .devcontainer/onCreateCommands.sh:28]: I got an error here (I believe):

```
[2023-09-12T13:56:39.662Z]     Building [====================>    ] 105/121: pyo3-macros, spin-macro, wit-bindgen-rust-impl, routefinder, pyo3-ffi                                                                
[2023-09-12T13:56:39.960Z] error: could not find native static library `python3.11`, perhaps an -L flag is missing?

   
[2023-09-12T13:56:39.960Z]  Building [====================>    ] 105/121: pyo3-macros, spin-macro, wit-bindgen-rust-impl, routefinder, pyo3-ffi                                                                
[2023-09-12T13:56:39.961Z]     Building [====================>    ] 106/121: pyo3-macros, memoffset, spin-macro, wit-bindgen-rust-impl, pyo3-ffi                                                                  
[2023-09-12T13:56:39.967Z] error: could not compile `pyo3-ffi` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
    Building [=====================>   ] 107/121: pyo3-macros, memoffset, spin-macro, wit-bindgen-rust-impl                                                                            
[2023-09-12T13:56:40.048Z]     Building [=====================>   ] 108/121: pyo3-macros, spin-macro, wit-bindgen-rust-impl                                                                                       
[2023-09-12T13:56:40.806Z]     Building [=====================>   ] 109/121: pyo3-macros, wit-bindgen-rust-impl                                                                                                   
[2023-09-12T13:56:43.246Z]     Building [=====================>   ] 110/121: pyo3-macros                                                                                                                          
[2023-09-12T13:56:43.825Z] 
[2023-09-12T13:56:43.879Z] make: *** [Makefile:17: /workspaces/spin-python-sdk/target/wasm32-wasi/release/spin_python_engine.wasm] Error 101
[2023-09-12T13:56:43.909Z] Stop (111755 ms): Run in container: /bin/sh -c .devcontainer/onCreateCommands.sh
[2023-09-12T13:56:43.910Z] onCreateCommand failed with exit code 2. Skipping any further user-provided commands.
[2023-09-12T13:56:43.912Z] Error: Command failed: /bin/sh -c .devcontainer/onCreateCommands.sh
[2023-09-12T13:56:43.912Z]     at TY (/Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js:235:130)
[2023-09-12T13:56:43.912Z]     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[2023-09-12T13:56:43.912Z]     at async Ml (/Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js:227:4393)
[2023-09-12T13:56:43.912Z]     at async Sl (/Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js:227:3738)
[2023-09-12T13:56:43.912Z]     at async kl (/Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js:227:2745)
[2023-09-12T13:56:43.912Z]     at async stA (/Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js:611:31548)
[2023-09-12T13:56:43.912Z]     at async otA (/Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js:611:28781)
[2023-09-12T13:56:43.916Z] Stop (112143 ms): Run: /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --ms-enable-electron-run-as-node /Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js run-user-commands --user-data-folder /Users/mikkel/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-session-data-folder /tmp/devcontainers-54ec42b8-c1a6-45c3-bf1f-a7927685e6901694526736464 --workspace-folder /Users/mikkel/code/fermyon/spin-python-sdk --id-label devcontainer.local_folder=/Users/mikkel/code/fermyon/spin-python-sdk --id-label devcontainer.config_file=/Users/mikkel/code/fermyon/spin-python-sdk/.devcontainer/devcontainer.json --container-id ca45dea2201f760a7b9b3ffa9ed3bcb490058009a61ce834c7454cef36bf53c9 --log-level debug --log-format json --config /Users/mikkel/code/fermyon/spin-python-sdk/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --skip-non-blocking-commands true --prebuild false --stop-for-personalization true --remote-env REMOTE_CONTAINERS_IPC=/tmp/vscode-remote-containers-ipc-3296e60d-496d-4136-852f-ded899109b69.sock --remote-env SSH_AUTH_SOCK=/tmp/vscode-ssh-auth-3296e60d-496d-4136-852f-ded899109b69.sock --remote-env REMOTE_CONTAINERS=true --mount-workspace-git-root true --terminal-columns 185 --terminal-rows 52 --dotfiles-target-path ~/dotfiles
[2023-09-12T13:56:43.916Z] Exit code 1
[2023-09-12T13:56:43.919Z] Command failed: /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --ms-enable-electron-run-as-node /Users/mikkel/.vscode/extensions/ms-vscode-remote.remote-containers-0.309.0/dist/spec-node/devContainersSpecCLI.js run-user-commands --user-data-folder /Users/mikkel/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-session-data-folder /tmp/devcontainers-54ec42b8-c1a6-45c3-bf1f-a7927685e6901694526736464 --workspace-folder /Users/mikkel/code/fermyon/spin-python-sdk --id-label devcontainer.local_folder=/Users/mikkel/code/fermyon/spin-python-sdk --id-label devcontainer.config_file=/Users/mikkel/code/fermyon/spin-python-sdk/.devcontainer/devcontainer.json --container-id ca45dea2201f760a7b9b3ffa9ed3bcb490058009a61ce834c7454cef36bf53c9 --log-level debug --log-format json --config /Users/mikkel/code/fermyon/spin-python-sdk/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --skip-non-blocking-commands true --prebuild false --stop-for-personalization true --remote-env REMOTE_CONTAINERS_IPC=/tmp/vscode-remote-containers-ipc-3296e60d-496d-4136-852f-ded899109b69.sock --remote-env SSH_AUTH_SOCK=/tmp/vscode-ssh-auth-3296e60d-496d-4136-852f-ded899109b69.sock --remote-env REMOTE_CONTAINERS=true --mount-workspace-git-root true --terminal-columns 185 --terminal-rows 52 --dotfiles-target-path ~/dotfiles
[2023-09-12T13:56:43.919Z] Exit code 1
```

**Review by mikkelhegn — APPROVED** [review]
LGTM

**Comment by mikkelhegn** [comment]
Thanks @jasonwashburn. One last thing is to to make sure your commits are signed before we can merge them.

**Comment by jasonwashburn** [comment]
I'm sure you guys are busy, just checking back to make sure there's not something additional you'd like to me to do with this?

**Comment by itowlson** [comment]
Sorry about that @jasonwashburn, I'm not sure what happened there!  I've merged it now - thanks for putting this together!

**Pull request merged**
